### PR TITLE
Part3: OpenIddict: Add Flows Event Handlers

### PR DESCRIPTION
**Pull request by Kukks**

### CODE REVIEW

**Comment by Kukks** [comment]
@NicolasDorier  rebased this now

File: BTCPayServer.Tests/AuthenticationTests.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                user.GrantAccess();
+
+                await GetAccessTokenByPasswordGrant(tester, user);
+            }
+        }
+
+        private static async Task<OpenIdConnectResponse> GetAccessTokenByPasswordGrant(ServerTester tester,
+            TestAccount user)
+        {
+            using (var httpClient = new HttpClient())
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/AuthenticationTests.cs]: use `tester.CreateClient()` or something like this.

**Comment by NicolasDorier** [comment]
Need rebase

**Comment by Kukks** [comment]
Rebased, I will be adding tests for each handler, at least as much as I can. Are there any specific tests you'd like to see for these @NicolasDorier ?

**Comment by NicolasDorier** [comment]
@Kukks yes, basically think about how we will document this stuff for people.

As far as I understand OpenId does not allow anything more, except if you specifically specify it at the controller level?

It would be nice for this PR to have at least 1 new API using this feature.
I propose the simplest one: `GetStatus` returning the `NBXplorerDashboard` infos.

**Comment by Kukks** [comment]
Yup, just keep in mind that even when this PR is merged in, there is still no way for users to create clients in order to make use of open id authorizations( that's in part 4)

**Comment by NicolasDorier** [comment]
@Kukks that's fine, if that's not implemented in UI yet, it is not needed for the tests.

**Comment by NicolasDorier** [comment]
@Kukks ha, I did not checked by can you make the Password flow disabled by default (then activable in the user's settings later?)

**Comment by NicolasDorier** [comment]
I will take a look, I don't like changing the default. One fuckup and people's server settings are open on the wide.

**Comment by Kukks** [comment]
I have a few more flows to test against and then I will improve on that dynamic policy default and make it less intrusive

**Comment by Kukks** [comment]
It now has tests for all flows (Password, Client Credentials, Implicit, Code, Refresh). When creating an API client, you need to specify which flows are allowed( so the password flow needs to be specifically chosen). I've had to re-add the dynamic policy because I think I found a .net core bug where the authorization filter priority is not consistent depending on where the attribute is defined in controllers but I have made it more restrictive( JWT bearer is only allowed when URL  starts with `/api` and has the `Authorization Bearer xxx` header)

**Comment by NicolasDorier** [comment]
I think you are mixing the concept of Policies with the concept of AuthenticationScheme. Will try to play around.

**Comment by NicolasDorier** [comment]
Ok what about this: You change the default authenticationScheme as one scheme that always fail.
That basically force people to specify clearly which scheme is needed everywhere, my last commit does that.

File: BTCPayServer.Tests/UnitTest1.cs [hunk 2]
```diff
@@ -230,9 +230,7 @@ public void CanCalculateCryptoDue()
 
             entity.Payments.Add(new PaymentEntity()
             {
-                Output = new TxOut(Money.Coins(0.5m), new Key()),
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/UnitTest1.cs]: Can you stop changing that back and forth? Before it was on one line, then you formatted to be on several line, and now you put back on one line.

File: BTCPayServer.Tests/AuthenticationTests.cs [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+        [Fact]
+        [Trait("Integration", "Integration")]
+        public void CanGetOpenIdConfiguration()
+        {
+            using (var tester = ServerTester.Create())
+            {
+                tester.Start();
+                var url = new Uri(tester.PayTester.ServerUri, "/.well-known/openid-configuration");
+                using (var wc = new WebClient())
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/AuthenticationTests.cs]: Use `tester.PayTester.HttpClient`

**Comment by NicolasDorier** [comment]
everything fail @kukks

**Comment by NicolasDorier** [comment]
Rebase,it was maybe my fault as I broke the build at one point.

File: BTCPayServer/Hosting/Startup.cs [hunk 4]
```diff
@@ -142,6 +163,10 @@ private void ConfigureOpenIddict(IServiceCollection services)
                 })
                 .AddServer(options =>
                 {
+                    if (_Env.IsDevelopment())
+                    {
+                        options.DisableHttpsRequirement();
```
Comment (NicolasDorier) [inline BTCPayServer/Hosting/Startup.cs]: mmmh what is requiring this? Because it will not work on Tor.

File: BTCPayServer.Tests/UnitTest1.cs [hunk 5]
```diff
@@ -253,20 +251,15 @@ public void CanCalculateCryptoDue()
...
-                Accounted = true,
-                NetworkFee = 0.1m
+                Output = new TxOut(Money.Coins(0.6m), new Key()), Accounted = true, NetworkFee = 0.1m
             });
 
             accounting = paymentMethod.Calculate();
             Assert.Equal(Money.Zero, accounting.Due);
             Assert.Equal(Money.Coins(1.3m), accounting.TotalDue);
 
-            entity.Payments.Add(new PaymentEntity()
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/UnitTest1.cs]: please remove those changes.

File: BTCPayServer/BTCPayServer.csproj [hunk 6]
```diff
@@ -1,6 +1,6 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Web">
-  <Import Project="../Build/Version.csproj" Condition="Exists('../Build/Version.csproj')" />
-  <Import Project="../Build/Common.csproj" />
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <Import Project="../Version.csproj" Condition="Exists('../Version.csproj')" />
```
Comment (NicolasDorier) [inline BTCPayServer/BTCPayServer.csproj]: revert this

**Comment by NicolasDorier** [comment]
tests still break, just rebase pelase

**Comment by Kukks** [comment]
Should be ok now

**Comment by NicolasDorier** [comment]
OK looking at it... again :D

**Comment by NicolasDorier** [comment]
Seems almost good!

* `AuthorizationCode_RefreshTokenGrantTypeEventHandler` should be splitted into two classes (or alternatively, put all the handlers into 1 class)
* Need to test `Logout`
* `AuthorizationEventHandler`, ` OpenIddictConstants.Errors.ServerError` is for bugs, should be InvalidGrant or something else.

**Comment by NicolasDorier** [comment]
tests not passing

**Comment by Kukks** [comment]
tests pass now :)

**Pull request merged**